### DISCUSSION TITLE
Anything needed for support of 'Arduino Nano' style RP2040 boards

### DISCUSSION THREAD

**Discussion by robzed** [tabemann/zeptoforth#38] [Ideas]
Hi, 

On UKMARSBot, some people are using 'Cytron Maker Nano RP2040' and 'Arduino Nano RP2040 Connect' boards. Do we need to do anything special to support these boards on Zeptoforth? I guess a minimal Pico Zeptoforth will work without the special features?

Boards:
* https://docs.arduino.cc/hardware/nano-rp2040-connect
* https://www.cytron.io/p-maker-nano-rp2040-simplifying-projects-with-raspberry-pi-rp2040

UKMARSBot https://ukmars.org/projects/ukmarsbot/ and https://github.com/ukmars

**Comment by tabemann** [comment]
zeptoforth itself should run on these as long as they use 25Q-type QSPI flash (both the Winbond and GigaDevices flash used in various RP2040-based boards qualify). I would presume that, most likely, such flash is being used. Some things such as LED's are not guaranteed to work out of the box (note that there is optional code included with zeptoforth which you can load which will enable it to use the Neopixels on the Cytron Maker Nano RP2040 ─ configuring it, though, is up to you). I should note that in particular the WiFi module on the Arduino Nano RP2040 Connect will not work out of the box (on that note I should buy one and get it working).

**Reply by tabemann** [reply]
There is one small caveat, though - GigaDevices 25Q flash is not good to build with. In theory, it *works*, but compiling to flash goes slowly. This should not be the matter with the Cytron Maker Nano RP2040 if they are telling the truth about having the same exact specs as the Pico (and using the same Winbond flash). However, if either of these uses GigaDevices 25Q flash, I would suggest doing significant builds to a Pico, then, pulling a UF2 file off the board with utils/download_uf2_image.sh, and then flashing the code onto the desired target.

Travis

**Reply by robzed** [reply]
The Arduino Nano RP2040 Connect uses an ISSI IS25LP128F-JKLE (NOR Flash 128M 2.3-3.6V 166Mhz Serial  - it's a 16Megabyte part) - I looked on the device.

The Cytron Maker Nano RP2040 I don't have (yet), so can't check the chip. The markings can be seen in this image: https://cdn.cnx-software.com/wp-content/uploads/2022/01/Maker-Nano-RP2040.jpg

**Reply by tabemann** [reply]
You could try the Arduino Nano RP2040 Connect - I can't guarantee that it will work, though.

**Reply by robzed** [reply]
Tried the Arduino Nano RP2040 Connect, using zeptoforth-0.64.2. (zeptoforth_full-0.64.2.bin)
Normal Forth interaction works. Tried this:

> compile-to-flash
> : hello1 ." hello rob" ;
> hello1 hello rob ok
> reboot
> hello1 hello rob ok
> <pull power, and replace>
> hello1 hello rob ok
> 

I wanted to check how much flash Zeptoforth thinks it's got attached... any ideas?

These don't appear to work on the RP2040 board?
free ( – ) Display space free for the main task and for flash in general
copyright ( – ) Display the copyright notices

**Reply by robzed** [reply]
What other tests do you want me to run on the Arduino Nano RP2040 Connect?

**Reply by tabemann** [reply]
My apologies - `free` and `copyright` don't exist. I had moved `free` to `unused` to avoid a name conflict with `free` used by the heap allocator, and I had merged the copyright notice with the license under `license`.

Thanks for confirming that it works on the Arduino Nano RP2040 Connect (even though it may take more work to get any LED's working ─ LED's have always been a trouble spot compatibility-wise). That's all I was really asking of you there - now we know that at at a basic level zeptoforth does work on the Arduino Nano RP2040 Connect.

About how much space zeptoforth thinks it has, it only stores compiled code in the lower 1 MB of flash. The second 1 MB of flash on the RP2040 is devoted to block storage. Any flash above that is currently ignored, but one could easily modify the code so it uses flash higher than 2 MB for more block storage.

**Comment by robzed** [comment]
I've created a demo for the Cytron Maker Nano RP2040 board. The code is located here: https://github.com/robzed/Maker_Nano_RP2040_Forth_Demo

I upload a video to Facebook. The information on this board is https://www.cytron.io/p-maker-nano-rp2040-simplifying-projects-with-raspberry-pi-rp2040

I'm happy that this board is working with Zeptoforth. (I have even compiled-to-flash and repowered - and it works).
